File: lib/api/models/system_setting.dart
// lib/models/system_setting.dart

class SystemSetting {
  final int? settingId;
  final String key;
  final Map<String, dynamic>? value;
  final bool isActive;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final int? updatedByUserId;
  final Map<String, dynamic>? updatedByUser;

  SystemSetting({
    this.settingId,
    required this.key,
    this.value,
    this.isActive = true,
    this.createdAt,
    this.updatedAt,
    this.updatedByUserId,
    this.updatedByUser,
  });

  factory SystemSetting.fromJson(Map<String, dynamic> json) {
    return SystemSetting(
      settingId: json['setting_id'],
      key: json['key'],
      value: json['value'] != null
          ? Map<String, dynamic>.from(json['value'])
          : null,
      isActive: json['is_active'] ?? true,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
      updatedByUserId: json['updated_by_user_id'],
      updatedByUser: json['updated_by_user'] != null
          ? Map<String, dynamic>.from(json['updated_by_user'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (settingId != null) 'setting_id': settingId,
      'key': key,
      'value': value,
      'is_active': isActive,
      if (createdAt != null) 'created_at': createdAt!.toIso8601String(),
      if (updatedAt != null) 'updated_at': updatedAt!.toIso8601String(),
      if (updatedByUserId != null) 'updated_by_user_id': updatedByUserId,
      if (updatedByUser != null) 'updated_by_user': updatedByUser,
    };
  }

  SystemSetting copyWith({
    int? settingId,
    String? key,
    Map<String, dynamic>? value,
    bool? isActive,
    DateTime? createdAt,
    DateTime? updatedAt,
    int? updatedByUserId,
    Map<String, dynamic>? updatedByUser,
  }) {
    return SystemSetting(
      settingId: settingId ?? this.settingId,
      key: key ?? this.key,
      value: value ?? this.value,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      updatedByUserId: updatedByUserId ?? this.updatedByUserId,
      updatedByUser: updatedByUser ?? this.updatedByUser,
    );
  }
}

File: final_project/models/system_setting.py
from datetime import datetime

from sqlalchemy import (
    Column, Integer, String, Boolean, DateTime, JSON, ForeignKey, Index, event
)
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship

from ..db import Base

class SystemSettingDB(Base):
    __tablename__ = "systemsetting"

    setting_id = Column(
        Integer,
        primary_key=True,
        autoincrement=True
    )  # Primary key
    key = Column(
        String(100),
        unique=True,
        nullable=False
    )  # Setting name/key
    value = Column(
        JSON,
        nullable=True
    )  # JSON value of the setting
    is_active = Column(
        Boolean,
        default=True,
        nullable=False
    )  # Whether the setting is active
    created_at = Column(
        DateTime,
        server_default=func.now(),
        nullable=False
    )
    updated_at = Column(
        DateTime,
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False
    )
    # Reference to the user who last updated this setting
    updated_by_user_id = Column(
        Integer,
        ForeignKey("users.user_id", ondelete="SET NULL"),
        nullable=True
    )

    __table_args__ = (
        Index("idx_systemsetting_is_active", "is_active"),
    )

    # Relationship to UserDB
    updated_by_user = relationship(
        "UserDB",
        back_populates="system_settings",
        foreign_keys=[updated_by_user_id]
    )

# Event listener to update 'updated_at' timestamp before updates
@event.listens_for(SystemSettingDB, 'before_update')
def _update_timestamp(mapper, connection, target):
    target.updated_at = datetime.utcnow()
